### DISCUSSION TITLE
Discrepancy in iCloud Drive listing

### DISCUSSION THREAD

**Discussion by EricEWeir** [mikekazakov/nimble-commander#137] [Q&A] [answered]
I get different directory lists for iCloud Drive in Finder and Nimble Commander. I want Nimble Commander to show what Finder shows. In th attached screenshot Nimble Commander is on the left, Finder on the right. Thanks in advance.

<img width="606" alt="Screenshot 2024-02-02 at 10 16 49 AM" src="https://github.com/mikekazakov/nimble-commander/assets/62710407/a31f2d9d-e7d9-4fb9-b722-685fa7227dd2">

**Comment by mikekazakov** [comment]
I can hazard to guess: NC shows what really exists on the file system, while Finder composes the items from multiple locations.
While it's not visible on the left window, likely the path is `~/Library/Mobile Documents/com~apple~CloudDocs`, which is a directory where iCloud Drive items are stored. Other application-specific folders (e.g. Pages or PDF Expert) are located on level up: `~/Library/Mobile Documents`.

**Reply by EricEWeir** [reply]
Thanks, Mike. Yes, I think all the missing folders are in `~/library/mobile documents`. A bit puzzled about why they are there instead of `~/library/mobile documents/com~apple~cloud docs`. I know at least one of particular interest was created in `icloud drive`, probably on an ipad.

But what I really want to know is whether there’s a way to create an `icloud drive` listing in NC that includes all the folders in the Finder listing.

**Reply by mikekazakov** [reply]
They are in different directories because semantically they belong to separate containers.
I don't think there is an official way to combine them.
While it's possible to "reverse engineer" what Finder does and replicate the same logic within NC, I'd be cautious in using that - there's no guarantee the setup won't suddenly break tomorrow with another update of MacOS.
Maybe it could be implemented as an experimental feature with that clearly stated...
Not sure how needed this functionality is, TBH.

If anybody has additional input - please chime in.

**Reply by EricEWeir** [reply]
thanks, mike. i was naive. i thought icloud drive was icloud drive. i’m discovering it’s not so simple. not clear how to proceed. maybe it’s not necessary that the directory with my obsidian vault appear in icloud drive, though on my mobile devices and in finder that’s where it is.

**Comment by WWWFWWW** [comment]
> thanks, mike. i was naive. i thought icloud drive was icloud drive. i’m discovering it’s not so simple. not clear how to proceed. maybe it’s not necessary that the directory with my obsidian vault appear in icloud drive, though on my mobile devices and in finder that’s where it is.

As I understand it (though I am no CloudKit developer!), the iCloud Drive is a bit different from CloudKit App's Documents (like your Obsidian files) as Mike wrote, even though both use the same iCloud storage space on Apple's servers and show up conflated in Finder or Drive on iOS.

It's a bit more detailed here:
https://developer.apple.com/library/archive/documentation/General/Conceptual/iCloudDesignGuide/Chapters/DesigningForDocumentsIniCloud.html

Not sure about the access thing, but it might be they even require the use of different APIs, as a CloudKit‑compatible app document saved in the iCloud might change at any time, since they can be accessed and changed on different devices at the same time, plus there is a bunch of special _methods_ to access them or push only the changed bits to the cloud, instead of the whole file again, to conserve bandwidth.

Some of the CloudKit apps might also present complex folders or bundles as single files, et cetera.

And of course there might be access issues as CloudKit might expect only the specific app in question (Obsidian) or Finder to access its container, even though you can get to it easily through NC or just terminal.

Again, I have zero experience in CloudKit dev., but to my laic eyes it feels better to be left alone. Just use the Finder to access the apps' data there, unfortunately.

**Reply by EricEWeir** [reply]
i'm responding to just this:

` ...to my laic eyes it feels better to be left alone. Just use the Finder to access the apps' data there....`

I think so. but I don't think I have to rely on finder. for starters, I've put my obsidian folder in favorites. I tried puttin a symlink to the obsidian folder in the icloud drive folder. but it seems I don't know how to do that. at least, if a link was created  I don't know where it went. 

help appreciated. or is it possible?

thanks

**Reply by EricEWeir** [reply]
i've been checking out commander one and forklift, too. the former has `/com~apple~clouddocs` which contains the same folders that nimble commander does. forklift has` /icloud drive` with the same set of folder that finder has.

**Reply by mikekazakov** [reply]
> I think so. but I don't think I have to rely on finder. for starters, I've put my obsidian folder in favorites. I tried puttin a symlink to the obsidian folder in the icloud drive folder. but it seems I don't know how to do that. at least, if a link was created I don't know where it went.
> 
> help appreciated. or is it possible?

I don't think that iCloud Drive would handle symlinks, but having a favourite bookmark to the specific app's container seems like a reasonable compromise to me.

**Reply by mikekazakov** [reply]
> i've been checking out commander one and forklift, too. the former has `/com~apple~clouddocs` which contains the same folders that nimble commander does. forklift has` /icloud drive` with the same set of folder that finder has.

I'm curious whether forklift has some secret sauce or they implemented this heuristic to combine some of the folders together.

**Reply by EricEWeir** [reply, marked as answer]
I dunno. (of course i don't. I'm just a user of tech  with limited knowledge of things below the surface.) i think I said forklift has an icloud drive directory containing the same folders that findger does. I misspoke. it originally had more folders than finder. I eliminated the ones I had no need for.

as far as I'm concerned, the issue is resolved. I've put my obsidian folder in favorites. and yesterday I noticed that appears in my root listing. and anyway, i don't have that much need to access either my icloud drive or obsidian folders from a file manager. only when setting up folders for use by apps.
